website/content/perf/2024-11-28-threading/index.md

```diff
@@ -0,0 +1,181 @@
++++
+title = "Measuring and Improving rustls's Multithreaded Performance"
+date = 2024-11-28
++++
+
+### System configuration
+
+We ran the benchmarks on a bare-metal server with the following characteristics:
+
+- OS: Debian 12 (Bookworm).
+- C/C++ toolchains: GCC 12.2.0 and Clang 14.0.6.
+- CPU: Ampere Altra Q80-30, 80 cores.
+- Memory: 128GB.
+- All cores set to `performance` CPU frequency governor.
+
+This is the [Hetzner RX170](https://www.hetzner.com/dedicated-rootserver/matrix-rx/).
+
+### How the benchmarks work
+Compared to [previous reports](/perf), the benchmark tool can now perform the same
+benchmarks in many threads simultaneously.  Each thread runs the same benchmarking
+operation as before, and threads do not contend with each other _except_ via
+the internals of the TLS library.
+
+As before, the benchmarking is performed by measuring a TLS client "connecting" to
+a TLS server over a memory buffer -- there is no network latency, system calls, or
+other overhead that would be present in a typical networked application.
+This arrangement actually should be the worst case for multithreaded testing: every
+thread should be working all the time (rather than waiting for IO) and therefore
+contention on any locks in the library under test should be maximal.
+
+### Versions
+The benchmarking tool used for both OpenSSL and BoringSSL was [openssl-bench `02249496`](https://github.com/ctz/openssl-bench/tree/02249496d2963e2a0b694e7be3b37f0d85f8eccd).
+
+#### BoringSSL and OpenSSL
+The version of BoringSSL and its build instructions are the same
+as [the previous report](2024-10-18-report.md).
```

Comment:
Because this is an index.md, I don't think this type of relative link works

Reply:
Doh, and our link checker just whinged about that. Will do a follow-up.

Reply:
That's why I noticed. I am not a markdown savant 😂 . (I'm excited to share these results)